File: php/api_key_client.php
<?php

include('httpful.phar');

function encode($data) {
    return str_replace(['+', '/'], ['-', '_'], base64_encode($data));
}

function decode($data) {
    return base64_decode(str_replace(['-', '_'], ['+', '/'], $data));
}

function genAuthHeaders($clientId, $apiClientKey, $dealerId, $userId, $apiKeyName, $apiKey, $method, $path) {
	// get the current date in yyyyMMdd format
	$date = date('Ymd');
	// current UTC timestamp in millisec
	$ts = round(microtime(true) * 1000);
	$ts = '1536353464539';

	$authHeader = "TCIv1-HmacSHA256 Credential=";
	$creds = array( $date, "external", $clientId, $dealerId, $userId, $apiKeyName );
	
	if( !isset( $dealerId ) || empty($dealerId)) { 
		$creds = array( $date, "external", $clientId, $userId, $apiKeyName );
	}
	
	$signingKey = $apiKey;
	
	// derive request signing key from the api key + creds
	for( $i = 0; $i < count($creds); $i++ ) {
		$signingKey = hash_hmac("sha256", $creds[$i], $signingKey, true);
		if ( $i > 0 ) {
			$authHeader = $authHeader . "/";
		}
		$authHeader = $authHeader . $creds[$i];
	}
	$authHeader = $authHeader . ", SignedEntities=METHOD;PATH;x-tci-timestamp, Signature=";

	// generate signature
	$toSign = $method . ";" . $path . ";" . $ts;
	$signature = hash_hmac('sha256', $toSign, $signingKey, true);
	$authHeader = $authHeader . bin2hex($signature);

	$headers = array (
		"x-api-key" => $apiClientKey,
		"Authorization" => $authHeader,
		"x-tci-timestamp" => $ts
	);

	return $headers;
} 

$clientId = $argv[1];
$apiClientKey = $argv[2];
$dealerId = $argv[3];
$userId = $argv[4];
$apiKeyName = $argv[5];
$apiKey = $argv[6];
$appId = $argv[7];

$method = "GET";
$path = "/applications/" . $appId;

echo($clientId . " " . $apiClientKey . " " . $dealerId . " " . $userId . " " . $apiKeyName . " " . $apiKey . " " . $method . " " . $path . "\n");

$headers = genAuthHeaders($clientId, $apiClientKey, $dealerId, $userId, $apiKeyName, $apiKey, $method, $path);

var_dump($headers);

$url = "https://api-demo.decisionlender.solutions" . $path;
$response = \Httpful\Request::get($url)
	->addHeaders($headers)
	->send();

echo($response);

?>

File: java/src/main/java/tci/dl4/restapi/client/ApiKeyClient.java
package tci.dl4.restapi.client;

import java.io.UnsupportedEncodingException;
import java.net.URLEncoder;
import java.text.SimpleDateFormat;
import java.util.Arrays;
import java.util.Date;
import java.util.List;
import java.util.Map.Entry;
import java.util.SortedMap;
import java.util.TreeMap;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;

import org.apache.commons.codec.binary.Hex;
import org.apache.commons.io.IOUtils;
import org.apache.http.client.methods.CloseableHttpResponse;
import org.apache.http.client.methods.HttpGet;
import org.apache.http.client.methods.HttpUriRequest;
import org.apache.http.impl.client.CloseableHttpClient;
import org.apache.http.impl.client.HttpClients;
import org.json.simple.JSONObject;
import org.json.simple.JSONValue;

public class ApiKeyClient {

	private static final SimpleDateFormat DATE_FORMATTER = new SimpleDateFormat("yyyyMMdd");
	private String basePath = null;
	private String apiClientId = null;
	private String apiUserName = null;
	private String apiKeyName = null;
	private SecretKeySpec initialKey;
	
	public ApiKeyClient(String basePath, String apiClientId, String apiUserName, String apiKeyName, String apiKeyValue)
			throws UnsupportedEncodingException {
		this.basePath = basePath;
		this.apiClientId = apiClientId;
		this.apiUserName = apiUserName;
		this.apiKeyName = apiKeyName;
		this.initialKey = new SecretKeySpec(apiKeyValue.getBytes("UTF-8"), "HmacSHA256");
	}

	private byte[] sign(String toSign, SigningKey signingKey) throws Exception {
		Mac hmacSHA256 = Mac.getInstance("HmacSHA256");
		hmacSHA256.init(signingKey.key);
		return hmacSHA256.doFinal(toSign.getBytes("UTF-8"));
	}
	
	// generate a signing key.  Current time must be passed as keys must rotate every 24 hours
	// The signing key is derived from the API Key generated in the UI along with the date,
	// client ID, user name and API key name
	private SigningKey keyGen(Date timestamp) throws Exception {
		String date = DATE_FORMATTER.format(timestamp);
		List<String> credentials = Arrays.asList( date, "external", apiClientId, apiUserName, apiKeyName );
		SigningKey signingKey = new SigningKey();
		signingKey.key = initialKey;

		String seperator = "";
		for ( String cred : credentials ) {
			byte[] signed = sign( cred, signingKey );
			signingKey.key = new SecretKeySpec(signed, "HmacSHA256");
			signingKey.credentials += seperator + URLEncoder.encode( cred, "UTF-8" );
			seperator = "/";
		}

		return signingKey;
	}
	
	// adds the HTTP Authorization header to the request.  the format is as follows:
	// TCIv1-HmacSHA256 Credential=yyyyMMdd/external/clientId/userId/keyName SignedEntities=METHOD,PATH,x-tci-timestamp Signature=<signatureInHex>
	// also adds a custom x-tci-timestamp header that is set to current time in millesconds since Unix epoch time
	protected void addAuthHeaderToRequest(HttpUriRequest req, String path) throws Exception {
		Date timestamp = new Date();
		SigningKey signingKey = keyGen(timestamp);
		
		String timestampStr = Long.toString(new Date().getTime());
		if ( ! path.startsWith("/") )
			path = "/" + path;
		
		SortedMap<String,String> toSignMap = new TreeMap<String,String>();
		toSignMap.put( "METHOD", req.getMethod() );
		toSignMap.put( "PATH", path );
		toSignMap.put( "x-tci-timestamp", timestampStr );
		
		String toSign = "";
		String toSignKeys = "";
		String seperator = "";
		for ( Entry<String,String> entry : toSignMap.entrySet() ) {
			toSignKeys += seperator + entry.getKey();
			toSign += seperator + entry.getValue();
			seperator = ";";
		}
		
		byte[] signatureBytes = sign( toSign, signingKey );
		String signature = Hex.encodeHexString(signatureBytes);
		
		String authHeader = "TCIv1-HmacSHA256";
		authHeader += " Credential=" + signingKey.credentials + ",";
		authHeader += " SignedEntities=" + toSignKeys + ","; 
		authHeader += " Signature=" + signature;

		req.setHeader("x-tci-timestamp", timestampStr);
		req.setHeader("Authorization", authHeader);
		
		System.out.println("Auth headers:");
		System.out.println("\tAuthorization -- " + authHeader);
		System.out.println("\tx-tci-timestamp -- " + timestampStr);
	}
	
	public JSONObject get(String path) throws Exception {
		HttpGet httpget = new HttpGet(basePath + path);
		addAuthHeaderToRequest(httpget, path);
		
		CloseableHttpClient httpclient = HttpClients.createDefault();
		CloseableHttpResponse response = httpclient.execute(httpget);
		String appContent = IOUtils.toString(response.getEntity().getContent());

		JSONObject jsonObject = null;
		if (appContent != null) {
			jsonObject = toJSONObject(appContent);
		}
		return jsonObject;
	}

	public JSONObject toJSONObject(String jsonStr) throws Exception {
		JSONObject jsonObj = null;
		if (jsonStr != null) {
			jsonObj = (JSONObject) JSONValue.parseWithException(jsonStr);
		}
		return jsonObj;
	}

	private static class SigningKey {
		String credentials = "";
		SecretKeySpec key = null;
	}

}

